features/step_definitions/iframe_steps.rb

```diff
@@ -12,12 +12,36 @@
   expect(@test_site.home).to have_index_iframe
 end
 
+Then(/^I can locate the iframe by name$/) do
+  @test_site.home.wait_for_named_iframe
+
+  expect(@test_site.home).to have_named_iframe
+end
+
+Then(/^I can locate the iframe by xpath$/) do
+  @test_site.home.wait_for_xpath_iframe
+
+  expect(@test_site.home).to have_xpath_iframe
+end
+
 Then(/^I can see elements in an iframe$/) do
   @test_site.home.my_iframe do |f|
     expect(f.some_text.text).to eq('Some text in an iframe')
   end
 end
 
+Then(/^I can see elements in a named iframe$/) do
```

Comment:
Can you refactor this step and all other similar steps (Including the ones that aren't yours), to remove the block. As it's superflous.

Here you would do: `@test_site.home.named_iframe.some_text.text`

Reply:
@luke-hill I tried without the block, but it does not work because as far as I know the block is necessary since Capybara needs to switch to the frame before being able to examine the elements within the frame:
```
  Scenario: locate an iframe by name         # features/iframes.feature:11
    When I navigate to the home page         # features/step_definitions/navigation_steps.rb:3
    Then I can locate the iframe by name     # features/step_definitions/iframe_steps.rb:15
    And I can see elements in a named iframe # features/step_definitions/iframe_steps.rb:33
      undefined method `call' for nil:NilClass
      Did you mean?  caller (NoMethodError)
      ./lib/site_prism/element_container.rb:60:in `block (2 levels) in iframe'
      ./lib/site_prism/element_container.rb:59:in `block in iframe'
      ./features/step_definitions/iframe_steps.rb:34:in `/^I can see elements in a named iframe$/'
      features/iframes.feature:14:in `And I can see elements in a named iframe'
```

As mentioned in the docs:

> Due to capybara internals it is necessary to pass a block to the iframe instead of simply calling methods on it; the block argument is the SitePrism::Page that represents the iframe's contents.

https://github.com/natritmeyer/site_prism/blame/master/README.md#L1557-L1561

Is there some other way I should try?

Reply:
Nope that's fine. I've not used iFrames much in my day-to-day. And the way I use it is with capybaras `within_iframe do` which uses a block. So it makes sense for the evaluation in SP to follow suit. Sorry for that.